go/parquet/file/column_writer.go

```diff
@@ -0,0 +1,567 @@
+	}
+	if pager.HasCompressor() {
+		ret.compressedTemp = new(bytes.Buffer)
+	}
+	if props.StatisticsEnabledFor(ret.descr.Path()) && ret.descr.SortOrder() != schema.SortUNKNOWN {
+		ret.pageStatistics = metadata.NewStatistics(ret.descr, props.Allocator())
+		ret.chunkStatistics = metadata.NewStatistics(ret.descr, props.Allocator())
+	}
+
+	if ret.props.DataPageVersion() == parquet.DataPageV1 {
+		if ret.descr.MaxDefinitionLevel() > 0 {
+			ret.defLevelSink.SetOffset(arrow.Uint32SizeBytes)
+		}
+		if ret.descr.MaxRepetitionLevel() > 0 {
+			ret.repLevelSink.SetOffset(arrow.Uint32SizeBytes)
+		}
+	}
+
+	ret.defEncoder.Init(parquet.Encodings.RLE, ret.descr.MaxDefinitionLevel(), ret.defLevelSink)
+	ret.repEncoder.Init(parquet.Encodings.RLE, ret.descr.MaxRepetitionLevel(), ret.repLevelSink)
+	return ret
+}
+
+func (w *columnWriter) SetBitsBuffer(buf *memory.Buffer) { w.bitsBuffer = buf }
+
+func (w *columnWriter) LevelInfo() LevelInfo { return w.levelInfo }
+
+func (w *columnWriter) Type() parquet.Type {
+	return w.descr.PhysicalType()
+}
+
+func (w *columnWriter) Descr() *schema.Column {
+	return w.descr
+}
+
+func (w *columnWriter) Properties() *parquet.WriterProperties {
+	return w.props
+}
+
+func (w *columnWriter) TotalCompressedBytes() int64 {
+	return w.totalCompressedBytes
+}
+
+func (w *columnWriter) TotalBytesWritten() int64 {
+	return w.totalBytesWritten
+}
+
+func (w *columnWriter) RowsWritten() int64 {
+	return int64(w.rowsWritten)
+}
+
+func (w *columnWriter) WriteDataPage(page DataPage) error {
+	written, err := w.pager.WriteDataPage(page)
+	w.totalBytesWritten += written
+	return err
+}
+
+func (w *columnWriter) WriteDefinitionLevels(levels []int16) {
+	w.defEncoder.EncodeNoFlush(levels)
+}
+
+func (w *columnWriter) WriteRepetitionLevels(levels []int16) {
+	w.repEncoder.EncodeNoFlush(levels)
+}
+
+func (w *columnWriter) init() {
+	w.defLevelSink.Reset(0)
+	w.repLevelSink.Reset(0)
+
+	if w.props.DataPageVersion() == parquet.DataPageV1 {
+		if w.descr.MaxDefinitionLevel() > 0 {
+			w.defLevelSink.SetOffset(arrow.Uint32SizeBytes)
+		}
+		if w.descr.MaxRepetitionLevel() > 0 {
+			w.repLevelSink.SetOffset(arrow.Uint32SizeBytes)
+		}
+	}
+
+	w.defEncoder.Reset(w.descr.MaxDefinitionLevel())
+	w.repEncoder.Reset(w.descr.MaxRepetitionLevel())
+}
+
+func (w *columnWriter) concatBuffers(defLevelsSize, repLevelsSize int64, values []byte, wr io.Writer) {
+	wr.Write(w.repLevelSink.Bytes()[:repLevelsSize])
+	wr.Write(w.defLevelSink.Bytes()[:defLevelsSize])
+	wr.Write(values)
+}
+
+func (w *columnWriter) EstimatedBufferedValueBytes() int64 {
+	return w.currentEncoder.EstimatedDataEncodedSize()
+}
+
+func (w *columnWriter) commitWriteAndCheckPageLimit(numLevels, numValues int64) error {
+	w.numBuffered += numLevels
+	w.numBufferedEncoded += numValues
+
+	if w.currentEncoder.EstimatedDataEncodedSize() >= w.props.DataPageSize() {
+		return w.AddDataPage()
+	}
+	return nil
+}
+
+func (w *columnWriter) AddDataPage() error {
```

Comment:
Would FlushCurrentPage be a better name for this method?

Reply:
renamed